From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 5 May 2025 13:04:14 +0000
Subject: test: force slow JSON.stringify path for overflow

Refs https://chromium-review.googlesource.com/c/v8/v8/+/6011806

The V8 team just enabled the JSON.stringify fast path by default,
which is iterative and won't throw. Pre-emptively preserve test
functionality by forcing the slow path. See above linked PR for
approach this is borrowed from.

Upstreamed in https://github.com/nodejs/node/pull/58181.

diff --git a/test/fixtures/console/stack_overflow.js b/test/fixtures/console/stack_overflow.js
index 565692b6d6e4ba4c439a38250407004c5aa21df9..14bceef878b5cf27609b29c8cc7852cab4dd63ff 100644
--- a/test/fixtures/console/stack_overflow.js
+++ b/test/fixtures/console/stack_overflow.js
@@ -26,11 +26,15 @@ Error.stackTraceLimit = 0;
 
 console.error('before');
 
+// Invalidate elements protector to force slow-path.
+// The fast-path of JSON.stringify is iterative and won't throw.
+Array.prototype[2] = 'foo';
+
 // Trigger stack overflow by stringifying a deeply nested array.
-let array = [];
-for (let i = 0; i < 100000; i++) {
-  array = [ array ];
-}
+// eslint-disable-next-line no-sparse-arrays
+let array = [,];
+for (let i = 0; i < 10000; i++)
+  array = [array];
 
 JSON.stringify(array);
 
